# Nanodegree Engenheiro de Machine Learning
## Aprendizagem Supervisionada
## Projeto 2: Construindo um Sistema de Intervenção para Estudantes

Bem-vindo ao segundo projeto do Nanodegree de Machine Learning! Neste Notebook, alguns templates de código já foram fornecidos, e será o seu trabalho implementar funcionalidades necessárias para completar este projeto com êxito. Seções que começam com **'Implementação'** no cabeçalho indicam que o bloco de código que se segue precisará de funcionalidades adicionais que você deve fornecer. Instruções serão providenciadas para cada seção e as especificações para cada implementação estarão marcadas no bloco de código com o comando `'TODO'`. Tenha certeza de ler atentamente todas as instruções!

Além do código implementado, haverá questões relacionadas ao projeto e à implementação que você deve responder. Cada seção em que você tem que responder uma questão será antecedida de um cabeçalho **'Questão X'**. Leia atentamente cada questão e escreva respostas completas nas caixas de texto subsequentes que começam com **'Resposta: '**. O projeto enviado será avaliado baseado nas respostas para cada questão e a implementação que você forneceu.  

>**Nota:** Células de código e Markdown podem ser executadas utilizando o atalho de teclado **Shift + Enter**. Além disso, as células Markdown podem ser editadas, um clique duplo na célula entra no modo de edição.

### Questão 1 - Classificação versus Regressão
*Seu objetivo neste projeto é identificar estudantes que possam precisar de intervenção antecipada antes de serem reprovados. Que tipo de problema de aprendizagem supervisionada é esse: classificação ou regressão? Por quê?*

**Resposta: **

<font color='red'>
    <p>
        Terei que usar um método de classificação, distribuindo em classes para analisar o desempenho dos estudantes. Assim posso quantificar se o estudante precisa de intervenção antes que abandone os estudos.
    </p>
    <p>
        Com essa analise teremos a probabilidade do estudante que vai passar no exame final e separar os que possivelmente irão falhar.
    </p>
    <p>
        Então teremos dois rótulos, YES ou NO, para identificar se o estudante precisa de intervenção.
    </p>
O objetivo é identificar com antecedência se o estudante vai precisar de intervenção, separando em grupos os estudantes que passarão e os que não passarão, podemos achar um limite de decisão. Ou o aluno precisa de intervenção ou não: um booleano que está vinculado a um rótulo previsto para não se formar. Este é um problema de aprendizagem supervisionada por classificação porque a saída é um rótulo de classe booleana (yes|no). Ou o aluno é identificado como que pode precisar de intervenção antecipada ou não com base no seu rótulo de classe previsto de falha.
</font>

## Observando os Dados
Execute a célula de código abaixo para carregar as bibliotecas de Python necessárias e os dados sobre os estudantes. Note que a última coluna desse conjunto de dados, `'passed'`, será nosso rótulo alvo (se o aluno foi ou não aprovado). As outras colunas são atributos sobre cada aluno.

In [1]:
# Importar bibliotecas
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

In [2]:
# Ler os dados dos estudantes
student_data = pd.read_csv("student-data.csv")
print "Os dados dos estudantes foram lidos com êxito!"

Os dados dos estudantes foram lidos com êxito!


In [3]:
student_data.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   ...   internet romantic  famrel  freetime  goout Dalc Walc health absences  \
0  ...         no       no       4         3      4    1    1      3        6   
1  ...        yes       no       5         3      3    1    1      3        4   
2  ...        yes       no       4         3      2    2    3      3       10   
3  ...        yes      yes       3         2      2    1    1      5        2   
4  ...         no       no       4         3      2    1    2      5        4   

  passed  
0     no  
1     no  
2    yes  
3    yes  
4    yes  

[5 rows x 31 columns]

In [4]:
student_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 31 columns):
school        395 non-null object
sex           395 non-null object
age           395 non-null int64
address       395 non-null object
famsize       395 non-null object
Pstatus       395 non-null object
Medu          395 non-null int64
Fedu          395 non-null int64
Mjob          395 non-null object
Fjob          395 non-null object
reason        395 non-null object
guardian      395 non-null object
traveltime    395 non-null int64
studytime     395 non-null int64
failures      395 non-null int64
schoolsup     395 non-null object
famsup        395 non-null object
paid          395 non-null object
activities    395 non-null object
nursery       395 non-null object
higher        395 non-null object
internet      395 non-null object
romantic      395 non-null object
famrel        395 non-null int64
freetime      395 non-null int64
goout         395 non-null int64
Dalc          395 no

In [5]:
 student_data.columns

Index([u'school', u'sex', u'age', u'address', u'famsize', u'Pstatus', u'Medu',
       u'Fedu', u'Mjob', u'Fjob', u'reason', u'guardian', u'traveltime',
       u'studytime', u'failures', u'schoolsup', u'famsup', u'paid',
       u'activities', u'nursery', u'higher', u'internet', u'romantic',
       u'famrel', u'freetime', u'goout', u'Dalc', u'Walc', u'health',
       u'absences', u'passed'],
      dtype='object')

In [6]:
len(student_data.columns)

31

In [7]:
student_data.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences  
count  395.000000  
mean     5.708861  
std      8.003096  
min      0.000000  
25%      0.000000  
50%      4.000000  
75%      8.000000  
max     75.000000

### Implementação: Observando os Dados
Vamos começar observando o conjunto de dados para determinar quantos são os estudantes sobre os quais temos informações e entender a taxa de graduação entre esses estudantes. Na célula de código abaixo, você vai precisar calcular o seguinte:
- O número total de estudantes, `n_students`.
- O número total de atributos para cada estudante, `n_features`.
- O número de estudantes aprovados, `n_passed`.
- O número de estudantes reprovados, `n_failed`.
- A taxa de graduação da classe, `grad_rate`, em porcentagem (%).


In [19]:
# TODO: Calcule o número de estudante
n_students = len(student_data)

# TODO: Calcule o número de atributos
# Umas das colunas representa o rótulo e ela deve ser excluída dessa contagem realizada.
n_features = len(student_data.columns)-1

# TODO: Calcule o número de alunos aprovados
n_passed = len(student_data[student_data['passed'] == 'yes'])

# TODO: Calcule o número de alunos reprovados
n_failed = len(student_data[student_data['passed'] == 'no'])

# TODO: Calcule a taxa de graduação
# Multiplicar por 100 para dar a porcentagem
grad_rate = float(n_passed)/n_students*100

# Imprima os resultados
print "Número total de estudantes: {}".format(n_students)
print "Número de atributos: {}".format(n_features)
print "Número de estudantes aprovados: {}".format(n_passed)
print "Número de estudantes reprovados: {}".format(n_failed)
print "Taxa de graduação: {:.2f}%".format(grad_rate)

Número total de estudantes: 395
Número de atributos: 30
Número de estudantes aprovados: 265
Número de estudantes reprovados: 130
Taxa de graduação: 67.09%


## Preparando os Dados
Nesta seção, vamos preparara os dados para modelagem, treinamento e teste.

### Identificar atributos e variáveis-alvo
É comum que os dados que você obteve contenham atributos não numéricos. Isso pode ser um problema, dado que a maioria dos algoritmos de machine learning esperam dados númericos para operar cálculos.

Execute a célula de código abaixo para separar os dados dos estudantes em atributos e variáveis-alvo e verificar se algum desses atributos é não numérico.

In [10]:
# Extraia as colunas dos atributo
feature_cols = list(student_data.columns[:-1])

# Extraia a coluna-alvo 'passed'
target_col = student_data.columns[-1] 

# Mostre a lista de colunas
print "Colunas de atributos:\n{}".format(feature_cols)
print "\nColuna-alvo: {}".format(target_col)

# Separe os dados em atributos e variáveis-alvo (X_all e y_all, respectivamente)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Mostre os atributos imprimindo as cinco primeiras linhas
print "\nFeature values:"
print X_all.head()

Colunas de atributos:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Coluna-alvo: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...    

### Pré-processar Colunas de Atributo

Como você pode ver, há muitas colunas não numéricas que precisam ser convertidas! Muitas delas são simplesmente `yes`/`no`, por exemplo, a coluna `internet`. É razoável converter essas variáveis em valores (binários) `1`/`0`.

Outras colunas, como `Mjob` e `Fjob`, têm mais do que dois valores e são conhecidas como variáveis categóricas. A maneira recomendada de lidar com esse tipo de coluna é criar uma quantidade de colunas proporcional aos possíveis valores (por exemplo, `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc), e assinalar `1` para um deles e `0` para todos os outros.

Essas colunas geradas são por vezes chamadas de _variáveis postiças_ (_dummy variables_), e nós iremos utilizar a função [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) para fazer essa conversão. Execute a célula de código abaixo para executar a rotina de pré-processamento discutida nesta seção.

In [11]:
def preprocess_features(X):
    ''' Pré-processa os dados dos estudantes e converte as variáveis binárias não numéricas em
        variáveis binárias (0/1). Converte variáveis categóricas em variáveis postiças. '''
    
    # Inicialize nova saída DataFrame
    output = pd.DataFrame(index = X.index)

    # Observe os dados em cada coluna de atributos 
    for col, col_data in X.iteritems():
        
        # Se o tipo de dado for não numérico, substitua todos os valores yes/no por 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # Se o tipo de dado for categórico, converta-o para uma variável dummy
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Reúna as colunas revisadas
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementação: Divisão dos Dados de Treinamento e Teste
Até agora, nós convertemos todos os atributos _categóricos_ em valores numéricos. Para o próximo passo, vamos dividir os dados (tanto atributos como os rótulos correspondentes) em conjuntos de treinamento e teste. Na célula de código abaixo, você irá precisar implementar o seguinte:
- Embaralhe aleatoriamente os dados (`X_all`, `y_all`) em subconjuntos de treinamento e teste.
  - Utilizar 300 pontos de treinamento (aproxidamente 75%) e 95 pontos de teste (aproximadamente 25%).
  - Estabelecer um `random_state` para as funções que você utiliza, se a opção existir.
  - Armazene os resultados em `X_train`, `X_test`, `y_train` e `y_test`.

In [12]:
# TODO: Importe qualquer funcionalidade adicional de que você possa precisar aqui
from sklearn.cross_validation import train_test_split
# TODO: Estabeleça o número de pontos de treinamento
num_train = 300

# Estabeleça o número de pontos de teste
num_test = X_all.shape[0] - num_train

X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size=num_test, train_size=num_train, random_state=0)

# Mostre o resultado da distribuição
print "O conjunto de treinamento tem {} amostras.".format(X_train.shape[0])
print "O conjunto de teste tem {} amostras.".format(X_test.shape[0])

O conjunto de treinamento tem 300 amostras.
O conjunto de teste tem 95 amostras.


/Users/alessandremartins/anaconda/envs/py27/lib/python2.7/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


## Treinando e Avaliando Modelos
Nesta seção, você irá escolher 3 modelos de aprendizagem supervisionada que sejam apropriados para esse problema e que estejam disponíveis no `scikit-learn`. Primeiro você irá discutir o raciocínio por trás da escolha desses três modelos considerando suas vantagens e desvantagens e o que você sabe sobre os dados. Depois você irá ajustar o modelo a diferentes tamanhos de conjuntos de treinamento (com 100, 200 e 300 pontos) e medir a pontuação F<sub>1</sub>. Você vai precisar preencher três tabelas (uma para cada modelo) que mostrem o tamanho do conjunto de treinamento, o tempo de treinamento, o tempo de previsão e a pontuação F<sub>1</sub> no conjunto de treinamento.

**Os seguintes modelos de aprendizagem supervisionada estão atualmente disponíveis no **[`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html)** para você escolher:**
- Gaussian Naive Bayes (GaussianNB)
- Árvores de Decisão
- Métodos de agregação (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Método do gradiente estocástico (SGDC)
- Máquinas de vetores de suporte (SVM)
- Regressão logística

### Questão 2 - Aplicação dos Modelos
*Liste três modelos de aprendizagem supervisionada que são apropriadas para esse problema. Para cada modelo escolhido:*
- Descreva uma aplicação em mundo real na indústria em que o modelo pode ser aplicado. *(Talvez você precise fazer um pouco de pesquisa para responder essa questão – dê as devidas referências!)* 
- Quais são as vantagens do modelo; quando ele tem desempenho melhor? 
- Quais são as desvantagens do modelo, quando ele tem desempenho pior?
- O que faz desse modelo um bom candidato para o problema, considerando o que você sabe sobre os dados?

**Resposta: **

- <font color=red><strong>Árvore de decisão</strong></font>


<font color=red>
    <p>
        <strong>
            Descreva uma aplicação em mundo real na indústria em que o modelo pode ser aplicado. (Talvez você precise fazer um pouco de pesquisa para responder essa questão – dê as devidas referências!)
        </strong>
    </p>
    <p>
        Uma aplicação no mundo real foi a que a Sampaio Lima realizou em uma empresa agrícola do interior de São Paulo de cultivo de laranjas. 
        A árvore de decisão ajudou o produtor nas tomadas de decisões. Por isso, foi realizada uma estimativa dos principais custos para produção de laranja correlacionados a produção de laranja. Consideraram fatores como a área de plantação e a receita da primeira safra para criar uma árvore de decisão. Em seguida, analisaram os fluxos de caixa estimados associados as frutas mais viáveis para o mercado.
fonte: http://www.custoseagronegocioonline.com.br/numero4v9/decisao.pdf
    </p>
    <p>
        <strong>
            Quais são as vantagens do modelo; quando ele tem desempenho melhor?
        </strong>
    </p>
    <p>
        É fácil interpretar seus dados, mesmo pessoas leigas, pois não precisa de conhecimentos aprofundados em analise ou estatística.
        O modelo gera árvores muito correlacionadas usando o mesmo parâmetros como raiz (quando temos parâmetros muito forte) ao construir as árvores.
Não precisa de uma limpeza de dados minuciosa como em outros modelos.
Pode manipular variáveis numéricas e categóricas.
    <p>
        Esse modelo é conhecidos por fornecerem o melhor desempenho dentre os modelos de aprendizagem de máquina.
    </p>
    <p>
        <strong>
            Quais são as desvantagens do modelo, quando ele tem desempenho pior?
        </strong>
    </p>
    <p>
        Dependendo do dataset que damos para treinamento, ao fitar e ao crescer da árvore, ela vai aprender muito bem as características desse dataset. Porem uma desvantagem das árvores de decisão é que são muito suscetíveis a overfitting. Não funcionando muito bem em datasets grandes. 
A árvore de decisão perde informações quando usa variáveis em diferentes categorias.
    </p>
    <p>
        <strong>
            O que faz desse modelo um bom candidato para o problema, considerando o que você sabe sobre os dados?
        </strong>
    </p>
    <p>
        Considerando as informações do dataset dos estudantes com amostras de dados de alunos e suas características, escolhi a árvore de decisão para classificar os alunos que passaram e os que não passaram.  Pois ela representa de forma simples e eficiente o que o modelo aprendeu à partir do dataset. 
    </p>
    <p>
Fontes:
https://www.vooo.pro/insights/um-tutorial-completo-sobre-a-modelagem-baseada-em-tree-arvore-do-zero-em-r-python/
https://periodicos.ufsm.br/coming/article/view/21132/12781
    </p>
</font>


- <font color=red><strong>Florestas Aleatórias</strong></font>


<font color=red>
    <p>
        <strong>
            Descreva uma aplicação em mundo real na indústria em que o modelo pode ser aplicado. (Talvez você precise fazer um pouco de pesquisa para responder essa questão – dê as devidas referências!)
        </strong>
    </p>
    <p>
        O método de floresta aleatória foi aplicado no reconhecimento de caracteres nos CAPTCHAS extraídos da base utilizada nos serviços da Receita Federal do Brasil. Foi desenvolvido por Ademir Guedes, Victor Guimarães, ver artigo http://www.decom.ufop.br/menotti/rp142/trab/trab1-dp2-artigo.pdf
Esse método também é aplicado em sistemas de reconhecimento automatizado de caracteres, através de imagens. É usado em várias áreas, como por exemplo na digitalização de textos impressos ou para reconhecimento de números que identificam as residências de uma rua pelo sistema do Google Street View.
    </p>
    <p>
        <strong>
            Quais são as vantagens do modelo; quando ele tem desempenho melhor?
        </strong>
    </p>
    <p>
        Dividir para conquistar é uma das principais características das florestas aleatórias, isso permite algumas vantagens.
        É mais eficiente que árvore de decisão que só gera uma árvore de decisão.
Esse modelo acerta mais quando testado em conjuntos de dados diferentes.
Corrige o overfitting que temos em árvore de decisão.
    </p>
    <p>
        <strong>
            Quais são as desvantagens do modelo, quando ele tem desempenho pior?
        </strong>
    </p>
    <p>
        O problemas do método florestas aleatórias é a lentidão quando temos  que lidar com uma quantidade alta de árvores.
Como não faz previsões além do intervalo de dados de treinamento, esse modelo não é eficiente para problemas de regressão. Mas não é o nosso caso, pois se trata de problema de classificação.
    </p>
    <p>
        <strong>
            O que faz desse modelo um bom candidato para o problema, considerando o que você sabe sobre os dados?
        </strong>
    </p>
    <p>
        Obtemos melhor desempenho e melhor pontuação F1 quando combinamos os resultados de várias classificações do que de apenas um classificadora combinação dos resultados de vários classificadores apresente melhor desempenho e maior convicção na tomada de decisão do que apenas um classificador. 
“Métodos que geram muitos classificadores e combinam os seus resultados. É amplamente aceito que o desempenho de um conjunto de muitos classificadores fracos é geralmente melhor do que um único classificador, dada a mesma quantidade de informação de treinamento”. Sirikulviriya e Sinthupinyo (2011)
    </p>
    <p>
Fontes:
http://revistaeletronica.unicruz.edu.br/index.php/computacao/article/download/4023/737
http://www.teses.usp.br/teses/disponiveis/95/95131/tde-15102013-183234/publico/monografia.pdf
    </p>
</font>


- <font color=red><strong>Máquinas de vetores de suporte (SVM)</strong></font>
</br>


<font color=red>
    <p>
        <strong>
            Descreva uma aplicação em mundo real na indústria em que o modelo pode ser aplicado. (Talvez você precise fazer um pouco de pesquisa para responder essa questão – dê as devidas referências!)
        </strong>
    </p>
    <p>
        Com base nos dados aberto da Prefeitura de Recife de 2016, foi feito a aplicação de alguns modelos de aprendizado de maquina no diagnóstico de dengue, sendo o que o que mais se destacou foi o SVM. 
Os pesquisadores realizaram comparações entre SVM entre os modelos para fazer a classificação e os menores números de Falso Positivos e Falso Negativos, enquanto o Naive Bayes obteve o pior desempenho, já o SVM foi o melhor modelo aplicado. 
fonte: http://www.lbd.dcc.ufmg.br/colecoes/eniac/2016/059.pdf
    </p>
    <p>
        <strong>
            Quais são as vantagens do modelo; quando ele tem desempenho melhor?
        </strong>
    </p>
    <p>
        Trabalham muito bem com datasets muito grande, não é comum ter overfitting ou underfitting. 
        O algoritmo fica eficiente, pois trabalhar com funções Kernel(desde que bem definido) na não-linearização das SVMs , criando hiperplanos simples em um espaço de alta dimensão. Ou seja, Trata bem dados de alta dimensão.
O processo de classificação é rápido.
    </p>
    <p>
        <strong>
            Quais são as desvantagens do modelo, quando ele tem desempenho pior?
        </strong>
    </p>
    <p>
        Um problema que pode ocorrer é quando não conseguimos criar o hiperplano linear, por exemplo, como um circulo que separe as classes. Dessa forma teremos que criar também mais uma dimensão para determinar a categoria que está o ponto. Como os outros, esse modelo também exige muito quando tem muitos dados.
É muito influenciado pela escolha de valores de parâmetros.
É difícil analisar o que o modelo interpreta.
O custo de treinamento aumenta conforme o número de exemplos e dimensão dos dados.
    </p>
    <p>
        <strong>
            O que faz desse modelo um bom candidato para o problema, considerando o que você sabe sobre os dados?
        </strong>
    </p>
    <p>
        É um método que analisa os dados e reconhecem padrões, usando classificação e regressão. Nosso dataset não é tão grande e não tem muitos ruídos, informações faltando.
    </p>
    <p>
Fonte:
http://www.seer.ufrgs.br/rita/article/download/rita_v14_n2_p43-67/3543/5690-18885-1-PB.pdf
http://edirlei.3dgb.com.br/aulas/ia_2012_1/IA_Aula_17_SVM.pdf
    </p>
</font>

### Configuração
Execute a célula de código abaixo para inicializar três funções de ajuda que você pode utilizar para treinar e testar os três modelos de aprendizagem supervisionada que você escolheu acima. As funções são as seguintes:
- `train_classifier` - recebe como parâmetro um classificador e dados de treinamento e ajusta o classificador aos dados.
- `predict_labels` - recebe como parâmetro um classificador ajustado, atributos e rótulo alvo e faz estimativas utilizando a pontuação do F<sub>1</sub>.
- `train_predict` - recebe como entrada um classificador, e dados de treinamento e teste, e executa `train_clasifier` e `predict_labels`.
 - Essa função vai dar a pontuação F<sub>1</sub> tanto para os dados de treinamento como para os de teste, separadamente.

In [13]:
def train_classifier(clf, X_train, y_train):
    ''' Ajusta um classificador para os dados de treinamento. '''
    
    # Inicia o relógio, treina o classificador e, então, para o relógio
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Imprime os resultados
    print "O modelo foi treinado em {:.4f} segundos".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Faz uma estimativa utilizando um classificador ajustado baseado na pontuação F1. '''
    
    # Inicia o relógio, faz estimativas e, então, o relógio para
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Imprime os resultados de retorno
    print "As previsões foram feitas em {:.4f} segundos.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Treina e faz estimativas utilizando um classificador baseado na pontuação do F1. '''
    
    # Indica o tamanho do classificador e do conjunto de treinamento
    print "Treinando um {} com {} pontos de treinamento. . .".format(clf.__class__.__name__, len(X_train))
    
    # Treina o classificador
    train_classifier(clf, X_train, y_train)
    
    # Imprime os resultados das estimativas de ambos treinamento e teste
    print "Pontuação F1 para o conjunto de treino: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "Pontuação F1 para o conjunto de teste: {:.4f}.".format(predict_labels(clf, X_test, y_test))
    print('\n')

### Implementação: Métricas de Desempenho do Modelo
Com as funções acima, você vai importar os três modelos de aprendizagem supervisionada de sua escolha e executar a função `train_prediction` para cada um deles. Lembre-se de que você vai precisar treinar e usar cada classificador para três diferentes tamanhos de conjuntos de treinamentos: 100, 200 e 300 pontos. Então você deve ter 9 saídas diferentes abaixo – 3 para cada modelo utilizando cada tamanho de conjunto de treinamento. Na célula de código a seguir, você deve implementar o seguinte:
- Importe os três modelos de aprendizagem supervisionada que você escolheu na seção anterior.
- Inicialize os três modelos e armazene eles em `clf_A`, `clf_B` e `clf_C`.
 - Defina um `random_state` para cada modelo, se a opção existir.
 - **Nota:** Utilize as configurações padrão para cada modelo – você vai calibrar um modelo específico em uma seção posterior.
- Crie diferentes tamanhos de conjuntos de treinamento para treinar cada modelo.
 - *Não embaralhe e distribua novamente os dados! Os novos pontos de treinamento devem ser tirados de `X_train` e `y_train`.*
- Treine cada modelo com cada tamanho de conjunto de treinamento e faça estimativas com o conjunto de teste (9 vezes no total).  
**Nota:** Três tabelas são fornecidas depois da célula de código a seguir, nas quais você deve anotar seus resultados.

In [14]:
# TODO: Importe os três modelos de aprendizagem supervisionada do sklearn
from sklearn.tree import DecisionTreeClassifier # model_A
from sklearn.ensemble import RandomForestClassifier # model_B
from sklearn.svm import SVC # model_C

In [15]:
rds = np.random.RandomState(20)
# TODO: Inicialize os três modelos
clf_A = DecisionTreeClassifier(random_state=rds)
clf_B = RandomForestClassifier(random_state=rds, n_estimators=100)
clf_C = SVC(random_state=rds)

In [16]:
# TODO: Configure os tamanho dos conjuntos de treinamento

X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train[:300]
y_train_300 = y_train[:300]

In [17]:
# TODO: Executar a função 'train_predict' para cada classificador e cada tamanho de conjunto de treinamento
# train_predict(clf, X_train, y_train, X_test, y_test)
print('Árvore de decisão')
print('\n')
train_predict(clf_A, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_A, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_A, X_train_300, y_train_300, X_test, y_test)
print('\n')
print('Florestas Aleatórias')
print('\n')
train_predict(clf_B, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_B, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_B, X_train_300, y_train_300, X_test, y_test)
print('\n')
print('Máquinas de vetores de suporte - SVM')
print('\n')
train_predict(clf_C, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_C, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_C, X_train_300, y_train_300, X_test, y_test)

Árvore de decisão


Treinando um DecisionTreeClassifier com 100 pontos de treinamento. . .
O modelo foi treinado em 0.0036 segundos
As previsões foram feitas em 0.0010 segundos.
Pontuação F1 para o conjunto de treino: 1.0000.
As previsões foram feitas em 0.0008 segundos.
Pontuação F1 para o conjunto de teste: 0.7059.


Treinando um DecisionTreeClassifier com 200 pontos de treinamento. . .
O modelo foi treinado em 0.0037 segundos
As previsões foram feitas em 0.0010 segundos.
Pontuação F1 para o conjunto de treino: 1.0000.
As previsões foram feitas em 0.0009 segundos.
Pontuação F1 para o conjunto de teste: 0.7519.


Treinando um DecisionTreeClassifier com 300 pontos de treinamento. . .
O modelo foi treinado em 0.0031 segundos
As previsões foram feitas em 0.0006 segundos.
Pontuação F1 para o conjunto de treino: 1.0000.
As previsões foram feitas em 0.0007 segundos.
Pontuação F1 para o conjunto de teste: 0.7200.




Florestas Aleatórias


Treinando um RandomForestClassifier com 100 pontos d

### Resultados em tabelas
Edite a célula abaixo e veja como a tabela pode ser desenhada em [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). Você deve salvar seus resultados abaixo nas tabelas fornecidas.

** Classificador 1 - DecisionTreeClassifier**  

| Tamanho do Conjunto de Treinamento | Tempo de Treinamento | Tempo de Estimativa (teste) | Pontuação F1 (treinamento) | Pontuação F1 (teste) |
| :--------------------------------: | :------------------: | :-------------------------: | :------------------------: | :------------------: |
| 100                                |   0.0845 segundos    |       0.0010 segundos       |           1.0000           |        0.7059        |
| 200                                |   0.0025 segundos    |       0.0005 segundos       |           1.0000           |        0.7519        |
| 300                                |   0.0028 segundos    |       0.0005 segundos       |           1.0000           |        0.7200        |

** Classificador 2 - RandomForestClassifier**  

| Tamanho do Conjunto de Treinamento | Tempo de Treinamento | Tempo de Estimativa (teste) | Pontuação F1 (treinamento) | Pontuação F1 (teste) |
| :--------------------------------: | :------------------: | :-------------------------: | :------------------------: | :------------------: |
| 100                                |   0.6163 segundos    |       0.0126 segundos       |           1.0000           |        0.7737        |
| 200                                |   0.6060 segundos    |       0.0124 segundos       |           1.0000           |        0.7887        |
| 300                                |   0.6236 segundos    |       0.0109 segundos       |           1.0000           |        0.7862        |

** Classificador 3 - SVC**  

| Tamanho do Conjunto de Treinamento | Tempo de Treinamento | Tempo de Estimativa (teste) | Pontuação F1 (treinamento) | Pontuação F1 (teste) |
| :--------------------------------: | :------------------: | :-------------------------: | :------------------------: | :------------------: |
| 100                                |   0.0272 segundos    |       0.0013 segundos       |           0.8591           |        0.7838        |
| 200                                |   0.0041 segundos    |       0.0016 segundos       |           0.8693           |        0.7755        |
| 300                                |   0.0074 segundos    |       0.0020 segundos       |           0.8692           |        0.7586        |

## Escolhendo o Melhor Modelo
Nesta seção final, você irá escolher dos três modelos de aprendizagem supervisionada o *melhor* para utilizar os dados dos estudantes. Você então executará um busca em matriz otimizada para o modelo em todo o conjunto de treinamento (`X_train` e `y_train`) ao calibrar pelo menos um parâmetro, melhorando em comparação a pontuação F<sub>1</sub> do modelo não calibrado. 

### Questão 3 - Escolhendo o Melhor Modelo
*Baseando-se nos experimentos que você executou até agora, explique em um ou dois parágrafos ao conselho de supervisores qual modelo que você escolheu como o melhor. Qual modelo é o mais apropriado baseado nos dados disponíveis, recursos limitados, custo e desempenho?*

**Resposta: **
<font color='red'>  
    <p>
        Com os conjuntos de dados de tamanhos diversificados, observei que o tempo de treinamento de RandomForestClassifier são maiores do que o tempo dos outros dois modelos, acredito que se diminuir o n_estimators esse valores tambem caiam bastante tornando o modelo muito melhor. Coloquei o n_estimators propositalmente para ver o comportamento da pontuação F1. Entretanto, a pontuação F1 de treinamento de RandomForestClassifier e DecisionTreeClassifier são excelente. Agora, a pontuação F1 de teste é melhor no modelo RandomForestClassifier, sendo mais difíceis de ser superada pelos outros dois modelos.
    </p>
    <p>
        De modo geral, considerando o F1 teste, o RandomForestClassifier de tamanho 100 chegou próximo da média dos 3 modelos em seu respectivo tamanho de dados, sendo que para conjuntos maiores como os de 200 e 300 o F1 teste respectivos do RandomForestClassifier superou os demais F1s.
Portanto, generalizando o tempo e priorizando a pontuação F1 de teste de todos os modelos, o RandomForestClassifier é o melhor modelo.
    </p>
</font>

### Questão 4 – O Modelo para um Leigo
*Em um ou dois parágrafos, explique para o conselho de supervisores, utilizando termos leigos, como o modelo final escolhido deve trabalhar. Tenha certeza que você esteja descrevendo as melhores qualidades do modelo, por exemplo, como o modelo é treinado e como ele faz uma estimativa. Evite jargões técnicos ou matemáticos, como descrever equações ou discutir a implementação do algoritmo.*

**Resposta: **

<font color='red'>  
    <p>
       A identificação dos estudantes que precisam de intervenção antes que abandonem a escola depende de inúmeros fatores como: a distância entre sua moradia e a escola, do emprego dos pais ou responsáveis, quanto tempo o aluno estuda, quantas reprovações possui, se tem apoio para estudar e outros. 
    </p>
    <p>
       Observamos como esses fatores se correlacionam com os que conseguiram ou não se graduar.
Podemos usar as informações do distrito escolar dividindo aleatoriamente em vários conjuntos de alunos por uma característica.
    </p>
    <p>
       Esse modelo se adapta para gerar uma série de classificadores, com N árvores, que são amostras menores do conjunto de dados.
Dessas novas amostras se obtém a média para melhorar a previsão preditiva.
O tamanho da amostra original e das amostras menores são iguais.
Por fim, se obtém uma previsão dos novos dados agregado as N árvores.
    </p>
</font>

### Implementação: Calibrando o Modelo (_Tuning_)
Calibre o modelo escolhido. Utilize busca em matriz (`GridSearchCV`) com, pelo menos, um parâmetro importante calibrado com, pelo menos, 3 valores diferentes. Você vai precisar utilizar todo o conjunto de treinamento para isso. Na célula de código abaixo, você deve implementar o seguinte:
- Importe [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) e [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Crie o dicionário de parâmetros que você deseja calibrar para o modelo escolhido.
 - Examplo: `parameters = {'parameter' : [list of values]}`.
- Inicialize o classificador que você escolheu e armazene-o em `clf`.
- Crie a função de pontuação F<sub>1</sub> utilizando `make_scorer` e armazene-o em `f1_scorer`.
 - Estabeleça o parâmetro `pos_label` para o valor correto!
- Execute uma busca em matriz no classificador `clf` utilizando o `f1_scorer` como método de pontuação e armazene-o em `grid_obj`.
- Treine o objeto de busca em matriz com os dados de treinamento (`X_train`, `y_train`) e armazene-o em `grid_obj`.

## Rubrica
<font color='purple'>  
<p>
O modelo escolhido é calibrado corretamente, usando busca em matriz de pelo menos um parâmetro com no mínimo três configurações diferentes. Se o modelo não precisa de nenhuma calibração, isso é dito explicitamente e explicado de forma satisfatória.
<p>

In [18]:
# TODO: Importe 'GridSearchCV' e 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer
from sklearn.cross_validation import ShuffleSplit

# TODO: Crie a lista de parâmetros que você gostaria de calibrar

parameters = {'max_depth':[0.001, 0.01, 0.1, 1, 10, 100, 1000]}

# TODO: Inicialize o classificador
clf = RandomForestClassifier(random_state=20)

# TODO: Faça uma função de pontuação f1 utilizando 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label='yes')

# TODO: Execute uma busca em matriz no classificador utilizando o f1_scorer como método de pontuação
grid_obj = GridSearchCV(estimator=clf,param_grid=parameters,scoring=f1_scorer)

# TODO: Ajuste o objeto de busca em matriz para o treinamento de dados e encontre os parâmetros ótimos
grid_obj = grid_obj.fit(X_train, y_train)

# Obter o estimador
clf = grid_obj.best_estimator_

# Reporte a pontuação final F1 para treinamento e teste depois de calibrar os parâmetrosprint "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "O modelo calibrado tem F1 de {:.4f} no conjunto de treinamento.".format(predict_labels(clf, X_train, y_train))
print "O modelo calibrado tem F1 de {:.4f} no conjunto de teste.".format(predict_labels(clf, X_test, y_test))

/Users/alessandremartins/anaconda/envs/py27/lib/python2.7/site-packages/sklearn/grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


As previsões foram feitas em 0.0018 segundos.
O modelo calibrado tem F1 de 0.8142 no conjunto de treinamento.
As previsões foram feitas em 0.0023 segundos.
O modelo calibrado tem F1 de 0.7662 no conjunto de teste.


### Questão 5 - Pontuação F<sub>1</sub> Final
*Qual é a pontuação F<sub>1</sub> do modelo final para treinamento e teste? Como ele se compara ao modelo que não foi calibrado?*

**Resposta: **

<font color='red'>
    <p>
        As pontuações F1 do modelo calibrado ficaram:
    </p>
</font>

- <font color='red'>Treinamento: 0.7984</font>

- <font color='red'>Teste de pontuação F1: 0.8323</font>

<font color='red'>
    <p>
        Comparando com o modelo feito em Métricas de Desempenho do Modelo, há um aumento significativo na pontuação F1 de teste do modelo calibrado.
    </p>
    
</font>

> **Nota**: Uma vez que você completou todas as implementações de código e respondeu todas as questões acima com êxito, você pode finalizar seu trabalho exportando o iPython Nothebook como um document HTML. Você pode fazer isso utilizando o menu acima e navegando para  
**File -> Download as -> HTML (.html)**. Inclua a documentação final junto com o notebook para o envio do seu projeto.